mage_ai/streaming/sources/nats_js.py

```diff
@@ -0,0 +1,175 @@
+                    ssl_ctx.load_verify_locations(self.config.ssl_config.cafile)
+                if self.config.ssl_config.certfile and self.config.ssl_config.keyfile:
+                    ssl_ctx.load_cert_chain(
+                        certfile=self.config.ssl_config.certfile,
+                        keyfile=self.config.ssl_config.keyfile
+                    )
+                connect_opts["tls"] = ssl_ctx
+
+            # Use NKEY if provided
+            if self.config.user_credentials:
+                connect_opts["user_credentials"] = self.config.user_credentials
+
+            # Establish connection with the configured options
+            self.nc = await nats.connect(**connect_opts)
+            self.js = self.nc.jetstream()
+
+            # Default consumer_name to stream_name if not provided
+            consumer_name = self.config.consumer_name or self.config.stream_name
+
+            await self.js.add_stream(name=self.config.stream_name, subjects=[self.config.subject])
+            self.psub = await self.js.pull_subscribe(self.config.subject, consumer_name)
+        except NoServersError as e:
+            self._print(f'Caught NoServersError while connecting to NATS server: {e}')
+        except Exception as e:
+            self._print(f'Caught exception while connecting to NATS server: {e}')
+
+    # Define callback methods
+    async def disconnected_cb(self):
+        self._print('Got disconnected!')
+
+    async def reconnected_cb(self):
+        self._print(f'Got reconnected to {self.nc.connected_url.netloc}')
+
+    async def error_cb(self, e):
+        self._print(f'There was an error: {e}')
+
+    async def closed_cb(self):
+        self._print('Connection is closed')
+
+    async def aclose_client(self):
+        await self.nc.close()
+
+    def init_client(self):
+        future = asyncio.run_coroutine_threadsafe(self.ainit_client(), self.loop)
+        future.result()
+
+    def close_client(self):
+        asyncio.run_coroutine_threadsafe(self.aclose_client(), self.loop)
+
+    def batch_read(self, handler: Callable):
+        self.init_client()
+
+        self._print("Starting batch read")
+
+        try:
+            while True:
+                messages = self.fetch_messages()
+                if not messages:
+                    continue
+                decoded_messages = [json.loads(msg.data.decode()) for msg in messages]
+                handler(decoded_messages)  # Pass the entire batch to the handler
+
+                # Acknowledge all messages in the batch
+                for message in messages:
+                    asyncio.run_coroutine_threadsafe(message.ack(), self.loop)
+        finally:
+            self.close_client()
+            self.stop_loop()
+
+    async def afetch_messages(self):
+        try:
+            msgs = await self.psub.fetch(self.config.batch_size, timeout=self.config.timeout)
+            # return [json.loads(msg.data.decode()) for msg in msgs]
```

Comment:
Given that JSON decoding is being done in both `read` and `batch_read`, it does appear to make sense to push it down here, no?

Reply:
Fixed 97e214d7b